File: cpp/example5.cpp
#include <iostream>
#include <cstdlib>
#include "exevalator.hpp"
#include "exevalator.cpp"

/**
 * The class providing a function available in expressions.
 */
class MyFun : public exevalator::ExevalatorFunctionInterface {

    /**
     * Invokes the function.
     * 
     * @param arguments The vector storing values of arguments.
     * @return double The return value of this function.
     */
    double operator()(const std::vector<double> &arguments) {
        if (arguments.size() != 2) {
            throw new exevalator::ExevalatorException("Incorrect number of args");
        }
        return arguments[0] + arguments[1];
    }
};

/**
 * An example to use the function defined above in an expression.
 */
int main() {

    // Create an instance of Exevalator Engine
    exevalator::Exevalator exevalator;

    try {

        // Connects the function available for using it in expressions
        exevalator.connect_function("fun", std::make_shared<MyFun>());

        // Evaluate the value of an expression
        double result = exevalator.eval("fun(1.2, 3.4)");

        // Display the result
        std::cout << "result: " << result << std::endl;

    } catch (exevalator::ExevalatorException &e) {
        std::cout << "Error occurred: " << e.what() << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

File: typescript/example5.ts
import Exevalator, { FunctionInterface, ExevalatorError } from "./exevalator";

/*
 * An example to create a function for available in expressions.
 */

/**
 * Function available in expressions.
 */
class MyFunction implements FunctionInterface {

    /**
     * Invoke the function.
     * 
     * @param arguments An array storing values of arguments.
     * @return The return value of the function.
     */
    public invoke(args: number[]): number {
        if (arguments.length != 2) {
            throw new ExevalatorError("Incorrected number of args");
        }
        return args[0] + args[1];
    }
}

// Create an instance of Exevalator Engine
let exevalator: Exevalator = new Exevalator();

// Connects the function available for using it in expressions
const fun: MyFunction = new MyFunction();
exevalator.connectFunction("fun", fun);

// Evaluate the value of an expression
const result: number = exevalator.eval("fun(1.2, 3.4)");

// Display the result
console.log(`result: ${result}`);
